designdocs/provider_contexts.md

```diff
@@ -0,0 +1,383 @@
+    }
+    ```
+
+3. New `ProviderToApplicationRegistryEntry` storage be initialized:
+
+    ```rust
+    // Alias for clarity
+    pub type ApplicationIndex = u16;
+
+    #[pallet::storage]
+    pub type ProviderToApplicationRegistryEntry<T: Config> = StorageDoubleMap<
+        _,
+        Twox64Concat,
+        ProviderId,
+        Twox64Concat,
+        ApplicationIndex,
+        ProviderRegistryEntry<T>,
+        OptionQuery,
+    >;
+    ```
+
+4. `MaxProviderNameSize` be increased to `256`.
+5. `MaxProviderLogo250X100Size` be created and the limit set to `131_072` (128 KiB).
+6. Introduce hash-based logo approval mechanism:
+    - Logos are not embedded during proposal submission. Instead, a content-addressed identifier (CID), such as an IPFS CID (using sha2-256), may be used to represent the logo image.
+    Governance must explicitly approve these logo hashes as part of the provider registration process.
+    When using a CID:
+    - The CID submitted during proposal must be reproducible from the logo bytes.
+    - The on-chain logic must validate the CID by recomputing it from the submitted image and comparing it with the one in the proposal.
+    - To do this, the encoding (e.g., raw vs dag-pb) and hashing (e.g., sha2-256) must be standardized or documented.
+
+    ```rust
+        /// ApprovedLogos is map of hash vs logo bytes.
+        /// This is used to store the approved logos for providers.
+        /// The key is CID of the logo image, and the value is the logo bytes.
+        /// This allows for easy retrieval of the logo by its hash
+        #[pallet:storage]
+        #[pallet::getter(fn approved_logos)]
+        pub ApprovedLogos: StorageMap<
+            _,
+            Blake2_128Concat,
+            LogoCID,
+            BoundedVec<u8, T::MaxProviderLogo250X100Size>,
+            OptionQuery
+        >;
+
+    ```
+
+### **Changes and additions in extrinsics** <a id='extrinsics'></a>
+
+1. The `propose_to_be_provider` extrinsic will now accept an optional list of hashes for images/logos to be approved by governance. The provider payload will include the `ProviderRegistryEntry` struct with the default name, localized names, and default logo hash.
+
+    ```rust
+
+        #[pallet::call_index(0)]
+        pub fn propose_to_be_provider(
+            origin: OriginFor<T>,
+            payload: ProviderRegistryEntry<T>,
+        ) -> DispatchResultWithPostInfo {
+            // Implementation details...
+        }
+    ```
+
+2. `propose_to_be_provider` to will insert hashes into the `ApprovedLogos` storage map.
+3. Introduce a new extrinsic `propose_to_add_application` which work in similar way to `propose_to_be_provider` but will be used for adding or updating application contexts. It will accept a `ProviderRegistryEntry` struct with the application name, localized names, and logo hashes.
+
+    ```rust
+        #[pallet::call_index(1)]
+        pub fn propose_to_add_application(
+            origin: OriginFor<T>,
```

Comment:
@wilwade should we separate add and update as separate extrinsics, assuming we can handle both with an optional application index 

On similar lines do we enable updating provider data via governance in same fashion?

Reply:
Hmm....

1. I think we can just make them set functions. Simplifies the setup.
2. Re-Provider Data. This makes me wonder again about an `Application 0` default instead of having Provider Data and Application data. Is there any other way we could unify these? Perhaps a `propose_to_set_provider_application` that takes an optional application id and if no application id it sets the provider level? Not sure if that makes it more confusing over having two extrinsics or not.

Reply:
We can see how that works, but we do want to be able to update the provider information as well so not sure it gets us far unless we just "don't" have provider information and only have application information.

Reply:
That's true, it's better if we can just follow same model on both applications and provider side for issuing updates that way it's more nuanced even with some extra code.

I will capture that in this document